### DISCUSSION TITLE
Cross-encoder and MTEB reranking tasks

### DISCUSSION THREAD

**Discussion by qiisziilbash** [embeddings-benchmark/mteb#1213] [Q&A] [answered]
I tried to use a cross-encoder model (e.g. `BAAI/bge-reranker-base`) with reranking tasks (e.g. `SciDocsRR`) and I was expecting the similarity scores to be computed directly from cross encoder using pair of (query, doc) but when I put a debugger and go into weeds of the code I can't see that but instead I see embeddings are still being computed and scores are computed from embeddings. 

What am I missing?
In general I wonder:
- How an embedding model and a reranker model is being treated differently for a given reranking task? 
- How a reranker model is used for retrieval tasks?

**Comment by KennethEnevoldsen** [comment]
@orionw I believe you are the one most familiar with this

**Reply by orionw** [reply, marked as answer]
Great question @qiisziilbash! Some [example code for cross-encoders is here](https://github.com/embeddings-benchmark/mteb?tab=readme-ov-file#using-a-cross-encoder-for-reranking). 

I think there could be two issues: 

(1) the cross-encoder needs initial results to rerank. If you don't have any, perhaps it's using an initial base embedder to get the first stage results.  You can skip this by providing the cross-encoder with `previous_results` path (see the example).  

The (2)nd issue could be that you doing a Reranking AbstractTask (I assume based on the name).  When adding the cross-encoder implementation I only tested it for retrieval -- not all tasks support cross-encoders (e.g. clustering can't, etc.).  Obviously reranking would also be a good fit for a cross-encoder.  It's quite possible that we're just missing something small to enable it for reranking tasks (see here for the [retrieval](https://github.com/embeddings-benchmark/mteb/blob/main/mteb/evaluation/evaluators/RetrievalEvaluator.py#L229) code and it seems like we'd just need to change a bit here in the [reranking code](https://github.com/embeddings-benchmark/mteb/blob/main/mteb/evaluation/evaluators/RerankingEvaluator.py#L376)).  Thanks for pointing this out, we'll make an issue to fix it.

**Comment by qiisziilbash** [comment]
Thank you so much @orionw 
I also have noticed in the [leaderboard](https://huggingface.co/spaces/mteb/leaderboard) if I only check cross-encoders there is no model shows up. but I see in one of your old comments [here](https://huggingface.co/spaces/mteb/leaderboard/discussions/102#663b81c7197afc0630dc053b) there are some models with that filter.  Is it the case that there is no cross encoder in the leaderboard anymore or there is an issue with that filter?

**Reply by orionw** [reply]
The only place we evaluated cross-encoders is what I did for the instruction-retrieval tab.  It's hard to put cross-encoders in the main tab since we average over tasks that cross-encoders can't do (bi-text mining, clustering, etc.). But it could be done for many of the other tabs.  

I'd love for someone who has more bandwidth to add some results on the main retrieval/reranking tab, I think it would be really interesting results and is understudied right now.

**Reply by qiisziilbash** [reply]
I see; one more question: what makes a model to show up in cross-encoder checkbox? when I go to instruction-retrieval tab and I check only cross encoders for example [this](https://huggingface.co/google/flan-t5-base) model shows up which I can't see its relationship to being a cross-encoder. would u help me understand that?

**Reply by orionw** [reply]
Great question! My understanding of a cross-encoder is a model that applies full attention over query and document to create a relevance score.  

I used Flan-T5 as a cross-encoder in a zero-shot MonoT5 fashion - giving it an instruction and asking it to determine the relevance (true/false) and then taking the softmax of the logits of the true/false tokens to determine the score. Flan-T5 is surprisingly good at this for not being trained for it.  In the [FollowIR paper](https://www.semanticscholar.org/paper/FollowIR%3A-Evaluating-and-Teaching-Information-to-Weller-Chang/77e07e5542b450a2ee3193993c552700ad9ba82d/figure/4) it was only few points worse than MonoT5 despite not training on any reranking data.

**Reply by KennethEnevoldsen** [reply]
We should probably consider adding a better UI for cross-encoders. I have seen others use a two-column approach, with the second being the cross-encoder. We could consider this an "experiment" (see #1211 )

**Reply by orionw** [reply]
What do you mean by “two column” @KennethEnevoldsen? Agree it could go well as an “experiment”!
